text/1004-transactions-api.rst

```diff
@@ -0,0 +1,619 @@
+
+**Deprecate** old transaction API::
+
+    DeprecationWarning: `connection.transaction()` is deprecated. Use
+    `pool.retry()` (preferred) or `connection.try_transaction()`.
+
+Add ``RetryOptions`` class:
+
+.. code-block:: Python
+
+    type DelayFn = (attempt: number) => number;
+
+    class AttempsOption(enum.Enum):
+        ALL = "ALL"
+        NETWORK_ERROR = "NETWORK_ERROR"
+        CONCURRENT_UPDATE = "CONCURRENT_UPDATE"
+        DEADLOCK = "DEADLOCK"
+
+    class RetryOptions:
+        def __init__(self, attempts=3): ...
+
+        def transaction_options(self,
+           isolation: str = None,
+           readonly: bool = None,
+           deferrable: bool = None,
+        ) -> RetryOptions: ...
+
+        def attempts(
+            which: AttemptsOption,
+            attempts: int,
+            delay_ms: Callable[[int], [float]],
+        ): RetryOptions: ...
+    }
+
+Introduce the abstract classes for queries:
+
+.. code-block:: python
+
+    class AsyncExecutor(abc.AbstraceBaseClass):
+        async def execute(self, query): ...
+        async def query(self, query: str, *args, **kwargs) -> datatypes.Set: ...
+        async def query_one(self, query: str, *args, **kwargs) -> typing.Any: ...
+        async def query_json(self, query: str, *args, **kwargs) -> str: ...
+        async def query_one_json(self, query: str, *args, **kwargs) -> str: ...
+
+    class Executor(abc.AbstractClass):
+        def query(self, query: str, *args, **kwargs) -> datatypes.Set: ...
+        def query_one(self, query: str, *args, **kwargs) -> typing.Any: ...
+        def query_json(self, query: str, *args, **kwargs) -> str: ...
+        def query_one_json(self, query: str, *args, **kwargs) -> str: ...
+        def execute(self, query: str) -> None: ...
+
+These base classes should be implemented by respective classes:
+
+.. code-block:: python
+
+    class AsyncIOTransaction(AsyncExecutor): ...
+    class AsyncIOConnection(AsyncExecutor): ...
+    class AsyncIOPool(AsyncExecutor): ...
+    class Transaction(Executor): ...
+    class Connection(Executor): ...
+    class Pool(Executor): ...
+
+
+Exceptions API
+``````````````
+
+Error hierarchy is amended by introducing ``TransientError``, ``NetworkError``
+and ``EarlyNetworkError`` with the following relationships:
+
+.. code-block:: python
+
+    class TransientError(TransactionError): ...
+    class TransactionDeadlockError(TransientError): ...
+    class TransactionSerializationError(TransientError): ...
+    class NetworkError(EdgeDBError): ...
+    class EarlyNetworkError(NetworkError): ...
+
+All network error within connection should be converted into
+``EarlyNetworkError`` or ``NetworkError``. Former is used in context
+where we catch network error before sending a request.
+
+
+Future Work
+===========
+
+Transaction on Specific Connection
+----------------------------------
+
+Do we want and how ``connection.retry()`` should work?
```

Comment:
There's probably a typo here, not sure I can understand the "do we want and how" part

Reply:
Improved wording a bit.